modules/nlu-testing/src/tools/visualisation.ts

```diff
@@ -0,0 +1,189 @@
+        y: v.map(o => pcaTest.predict([o.utt_emb]).get(0, 1)),
+        z: v.map(o => pcaTest.predict([o.utt_emb]).get(0, 2)),
+        mode: 'markers',
+        type: 'scatter3d',
+        name: `${k}_test`,
+        text: v.map(o => o.utt),
+        marker: { size: 8, color: i }
+      })
+    )
+  }
+  return traces
+}
+
+export async function computeKmeansPairwiseIntent(state: BotState) {
+  const grouped_intents = _.groupBy(state.trainDatas, 'intent')
+  const simMat = { matrix: [], labels: [], text: [] }
+  for (const [intent, o] of Object.entries(grouped_intents)) {
+    const rowMat = []
+    const rowText = []
+    for (const [intentTodo, oTodo] of Object.entries(grouped_intents)) {
+      if (intent === intentTodo) {
+        rowMat.push(0)
+        rowText.push('')
+      } else {
+        let bestBadUttsText = []
+        let bestBadUttsNb = 10000
+        for (let i = 0; i < 10; i++) {
+          const kmeans = new clustering.KMEANS()
+          const [cluster1, cluster2] = kmeans.run(
+            o.concat(oTodo).map(o => o.utt_emb),
+            2
+          )
+          const clusterO = _.mean(cluster1) < _.mean(cluster2) ? cluster1 : cluster2
+          const badUttsText = []
+          let badUttsNb = 0
+          o.map((p, i) => {
+            if (!clusterO.includes(i)) {
+              badUttsText.push(p.utt)
+              badUttsNb += 1
+            }
+          })
+          if (badUttsNb < bestBadUttsNb) {
+            bestBadUttsNb = badUttsNb
+            bestBadUttsText = badUttsText
+          }
+        }
+        rowMat.push(bestBadUttsNb)
+        rowText.push(bestBadUttsText.join('<br>'))
+      }
+    }
+    simMat.labels.push(intent)
+    simMat.matrix.push(rowMat)
+    simMat.text.push(rowText)
+  }
+  const plotlyMatrixData = [
+    {
+      x: simMat.labels,
+      y: simMat.labels,
+      z: simMat.matrix,
+      text: simMat.text,
+      type: 'heatmap'
+    }
+  ]
+  return plotlyMatrixData
+}
+
+function closest(a: number[], b: number[][], index: number): number {
+  let minIndex = undefined
+  let minDistance = 100000
+  b.map((point, i) => {
+    const dist = distance.euclidean(a, point)
+    if (dist < minDistance && i !== index) {
+      minIndex = i
+      minDistance = dist
+    }
+  })
+  return minDistance
+}
+
+export function computeOutliers(state: BotState) {
+  const intentsData = _.groupBy(state.trainDatas, 'intent')
+  const dbPerIntent = _.mapValues(intentsData, o => {
+    const embedArray = o.map(e => e.utt_emb)
+    const meanDist = o.reduce((sum, curr, index) => {
+      if (index < o.length - 1) {
+        return sum + closest(curr.utt_emb, embedArray, index)
+      }
+      return sum / o.length
+    }, 0)
+
+    const dbscan = new clustering.DBSCAN()
+    // parameters: 5 - neighborhood radius, 2 - number of points in neighborhood to form a cluster
+    const clusters = dbscan.run(embedArray, meanDist + 0.5 * meanDist, _.floor(o.length / 3))
```

Comment:
What does this comment refer to ? are there some magic numbers somewhere ?

Reply:
Just to remember the parameters, because the library don't give doc or typings

Reply:
Nothing stops you from creating the typings yourself. Ex:
```js
interface DbScan {
  run: (neighboorhoodRadius: number, numberOfPoints: number) => number
}

const dbscan: DbScan = new clustering.DBSCAN()
```

and you can add other methods too, then you have autocompletion